### PR TITLE
Create pytorch_lightning/utilities/types.py

### PR DESCRIPTION
**Pull request by carmocca**
## What does this PR do?

File with common types that can be used throughout the codebase.

The current list is not exhaustive, just meant as a starting point for other PRs to use.

## Before submitting
- [x] Was this discussed/approved via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your PR does only one thing, instead of bundling different changes together?
- [n/a] Did you make sure to update the documentation with your changes? (if necessary)
- [n/a] Did you write any new necessary tests? (not for typos and docs)
- [x] Did you verify new and existing tests pass locally with your changes?
- [n/a] Did you update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review

 - [x] Is this pull request ready for review? (if not, please submit in draft mode)
 - [x] Check that all items from **Before submitting** are resolved
 - [x] Make sure the title is self-explanatory and the description concisely explains the PR
 - [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

### CODE REVIEW

**Review by ananthsub — APPROVED** [review]

File: pytorch_lightning/utilities/types.py [hunk 1]
```diff
@@ -0,0 +1,8 @@
+from typing import Dict, Iterator, Optional, Union
+
+import torch
+from torchmetrics import Metric
+
+_METRIC = Union[Metric, torch.Tensor, int, float]
+_STEP_OUTPUT = Optional[Union[torch.Tensor, Dict[str, torch.Tensor]]]
+_PARAMETERS = Iterator[torch.nn.Parameter]
```
Comment (Borda) [inline pytorch_lightning/utilities/types.py:8]:
```suggestion
_HINT_METRIC = Union[Metric, torch.Tensor, int, float]
_HINT_STEP_OUTPUT = Optional[Union[torch.Tensor, Dict[str, torch.Tensor]]]
_HINT_PARAMETERS = Iterator[torch.nn.Parameter]
```
or keep typing just to keep clear from the first look that it typing Constants...

File: pytorch_lightning/utilities/types.py [hunk 2]
```diff
@@ -0,0 +1,8 @@
+from typing import Dict, Iterator, Optional, Union
+
+import torch
+from torchmetrics import Metric
+
+_METRIC = Union[Metric, torch.Tensor, int, float]
```
Comment (awaelchli) [inline pytorch_lightning/utilities/types.py:6]: is it necessary to put the underscore?

File: pytorch_lightning/utilities/types.py [hunk 3]
```diff
@@ -0,0 +1,8 @@
...
+_STEP_OUTPUT = Optional[Union[torch.Tensor, Dict[str, torch.Tensor]]]
```
Comment (awaelchli) [inline pytorch_lightning/utilities/types.py:7]: step outputs can be anything inside the dict, it's not restricted to tensor, right?

File: pytorch_lightning/utilities/types.py (hunk 1 above)
Comment (carmocca) [reply pytorch_lightning/utilities/types.py:8]: Well, adding hint seems way too verbose

And keep in mind you do `from typing import Any`, not `from typing import AnyType`

File: pytorch_lightning/utilities/types.py (hunk 2 above)
Comment (carmocca) [reply pytorch_lightning/utilities/types.py:6]: Do we want users to import these and care about backwards compatibility?

**Review by tchaton — APPROVED** [review]
LGTM ! Let s update step_output

File: pytorch_lightning/utilities/types.py (hunk 1 above)
Comment (Borda) [reply pytorch_lightning/utilities/types.py:8]: yes I see but how many times are you scrolling up to see whom where it is imported...
on the other hand we have some other func names too long and this is just not clear at the fist look
so lets stay with word TYPES in the var name?

Comment (carmocca) [reply pytorch_lightning/utilities/types.py:8]: I personally prefer no suffixes. It should be obvious that it's a type considering its placement (`: METRIC` or `-> METRIC:`)

🎉 : `_TYPE` suffix (`_METRIC_TYPE`)
🚀 : No suffix (`_METRIC`)

cc: @PyTorchLightning/core-contributors for opinions as this will impact all codebase

File: pytorch_lightning/utilities/types.py (hunk 2 above)
Comment (Borda) [reply pytorch_lightning/utilities/types.py:6]: yes, as it is protected - just internal usage
BUT for step types, it would make sense to make it public so users can reuse our typing
cc: @ananthsub

File: pytorch_lightning/utilities/types.py (hunk 1 above)
Comment (Borda) [reply pytorch_lightning/utilities/types.py]: in the context of exposing some to the public like step types, it could be better to have the TYPE sufix

**Review by s-rog — APPROVED** [review]
